crates/nu-parser/src/parser.rs

```diff
@@ -608,6 +608,7 @@ fn calculate_end_span(
                 spans.len()
             } else if positional_idx < signature.required_positional.len()
                 && spans.len() > (signature.required_positional.len() - positional_idx)
+                && signature.category != Category::Default
```

Comment:
If we allow custom commands to change their category in the future, this might break. Is there a way we can narrow it down to where the command parsing is being impacted? or improve the logic so we test that we have enough spans ahead of time?

Reply:
@jntrnr Let us take this statement as an example `if 2 < 3 { 3 }`. 
- This comes with this signature: `signature Signature { name: "if", usage: "", extra_usage: "", search_terms: [], required_positional: [PositionalArg { name: "cond", desc: "condition to check", shape: Expression, var_id: None, default_value: None }, PositionalArg { name: "thenOf([Block, Expression])), var_id: None, default_value: None }], rest_positional: None, vectorizes_over_list: false, named: [Flag { long: "help", short: Some('h'), arg: None, required: false, desc: "Display the help message for this command", var_id: None, default_value: None }], input_type: Any, output_type: Any, input_output_types: [(Any, Any)], allow_variants_without_examples: false, is_filter: false, creates_scope: false, allows_unknown_args: false, category: Core }`
- And this span array: `[Span { start: 21046, end: 21047 }, Span { start: 21048, end: 21049 }, Span { start: 21050, end: 21051 }, Span { start: 21052, end: 21057 }]`

The first positional argument corresponds to `2 < 3` and so it includes 3 span elements but the second is a block `{ 3 }` which contains just 1 span element. And this is what we should be able to know: the number of span elements corresponding to each positional argument, knowing that this can vary as a function of the input statement from the user. But I couldn’t find any function or utility that links the two, do you have an idea on the issue?

Reply:
I'm not sure if I follow what the issue is. If the issue is that the arguments number isn't being properly counted, changing the end span calculation isn't where you'd want to do the fix.

I'm mostly saying this also because that part of the parser is also pretty nasty to get right. If the fix can be done elsewhere, that'll be a lot easier to test and a lot easier to review.